# Object Oriented Programming Principles Exercises

## Write codes based on the questions
---

### Classes Recap

**Qn 1)** Create a deck of cards class. Internally, the deck of cards should use another class, a `Card` class. Your requirements are:

* The `Deck` class should have a `deal` method to deal a single card from the deck
* After a card is dealt, it is removed from the deck.
* There should be a `shuffle` method which makes sure the deck of cards has all 52 cards and then rearranges them randomly.
* The `Card` class should have a suit (Hearts, Diamonds, Clubs, Spades) and a value (A,2,3,4,5,6,7,8,9,10,J,Q,K)
* use the `shuffle()` function from the `random` module to help you shuffle the cards

The test program have been provided.

**Sample Program Output**
<pre>
2 of Spades
3 of Diamonds
8 of Diamonds
J of Diamonds
7 of Hearts
Cards remaining in deck: 47
</pre>

In [ ]:
# test program -------------------------
d = Deck()
d.shuffle()

for _ in range(5):
    print(d.deal())
print(d)

---
### Encapsulation, Inheritance, Abstraction & Polymorphism

**Qn 1)** Someone made the following class:
```python
class Address:
    def __init__(self, street, num):
        self.street_name = street
        self.number = num
```

Sally now wants to make a subclass (child) of the class `Address` called `CampusAddress` that has a new attribute (`office number`),
that can vary. This subclass will always have the `street` attribute set to `Hougang` and the `num` attribute
set to `77`. She wants to use the class as follows:

```python
sally_addr = CampusAddress('8745 5951')
print(sally_addr.office_number)
print(sally_addr.street_name)
print(sally_addr.number)
```

Help her implement the `CampusAddress` class as she doesn't know how.

In [105]:
class Address:
    def __init__(self, street, num):
        self.street_name = street
        self.number = num

In [106]:
class CampusAddress(Address):
    
    def __init__(self, office_number, street = 'Hougang', num = 77):
        super().__init__(street, num)
        self.office_number = office_number

In [107]:
sally_addr = CampusAddress('8745 5951')
print(sally_addr.office_number)
print(sally_addr.street_name)
print(sally_addr.number)

8745 5951
Hougang
77


**Qn 2)** Write an abstract class (you are to decide whether this class is an Abstract Class or an Interface), `Box`, define the following methods: 
* `add` - for adding any number of items to the box
* `empty` - for taking all the items out of the box and returning them as a list
* `count` - for counting the items which are currently in the box. 

Write an `Item` class which has a `name` attribute and a `value` attribute. Assume that all the items you will use will be `Item` objects. Now write two more subclasses of `Box` which uses different underlying collections to store items: `ListBox` should use a `list`, and `DictBox` should use a `dict`.

Write a function called `repack_boxes`, which takes a list of boxes as parameters, gathers up all the items they contain, and redistributes them as evenly as possible over all the boxes. **Order is unimportant**.

**Sample Program Output**
<pre>
<b>Input:</b> 20 `ListBox` items, 9 `ListBox` items and 5 `DictBox` with 5 items
<b>Output:</b> 
12
11
11
</pre>

In [136]:
from abc import *

class Box(ABC):
    
    @abstractmethod
    def add(item, box):
        pass
    
    @abstractmethod
    def empty(box):
        pass
    
    @abstractmethod
    def count(box):
        pass
    
class ListBox(Box):
    
    def __init__(self):
        self.items = []
        
    def add(self, item):
        self.items.append(item)
        
    def empty(self):
        listOfItems = self.items
        self.items = []
        return listOfItems
    
    def count(self):
        return len(self.items)
    
class DictBox(Box):
    
    def __init__(self):
        self.items = {}
        
    def add(self, item):
        self.items.update({item: 0})
        
    def empty(self):
        listOfItems = [key for key in self.items]
        self.items = {}
        return listOfItems
    
    def count(self):
        return len(self.items)
    
class Item:
    
    def __init__(self, name, value):
        self._name = name
        self._value = value

In [177]:
def repack_boxes(lst):
    
    lstItems = list()
    
    for box in lst:
        boxContent = box.empty()
#         print("Length box content: ", len(boxContent))
#         print("Box.count() is now: ", box.count())
#         print("Length of lst is now: ", len(lst))
        for item in boxContent:
            # print("item is: ", item)
            lstItems.append(item)
    
    print("len(lstItems) is: ",len(lstItems))
    
    while lstItems:
        print("before for-loop, len(lstItems) is: ", len(lstItems))
        if len(lstItems) == 0:
            break
                
        for box in lst:
            print("len(lstItems): ", len(lstItems))
            box.add(lstItems.pop())
#             try:
#                 box.add(lstItems.pop())
#             except IndexError:
#                 break
            
            
            
    return None
    
    

In [178]:
# Test program ----------------------------------------
box1 = ListBox()
for i in range(20):
    box1.add(Item('box1, ' + str(i), i))
print("box1.count() now is: ", box1.count())
print()
box2 = ListBox()
for i in range(9):
    box2.add(Item('box2, ' + str(i), i))
print("box2.count() now is: ", box2.count())
print()
box3 = DictBox()
for i in range(5):
    box3.add(Item('box3, ' + str(i), i))
print("box3.count() now is: ", box3.count())
print()
repack_boxes([box1, box2, box3])

print(box1.count())
print(box2.count())
print(box3.count())

box1.count() now is:  20

box2.count() now is:  9

box3.count() now is:  5

len(lstItems) is:  34
before for-loop, len(lstItems) is:  34
len(lstItems):  34
len(lstItems):  33
len(lstItems):  32
before for-loop, len(lstItems) is:  31
len(lstItems):  31
len(lstItems):  30
len(lstItems):  29
before for-loop, len(lstItems) is:  28
len(lstItems):  28
len(lstItems):  27
len(lstItems):  26
before for-loop, len(lstItems) is:  25
len(lstItems):  25
len(lstItems):  24
len(lstItems):  23
before for-loop, len(lstItems) is:  22
len(lstItems):  22
len(lstItems):  21
len(lstItems):  20
before for-loop, len(lstItems) is:  19
len(lstItems):  19
len(lstItems):  18
len(lstItems):  17
before for-loop, len(lstItems) is:  16
len(lstItems):  16
len(lstItems):  15
len(lstItems):  14
before for-loop, len(lstItems) is:  13
len(lstItems):  13
len(lstItems):  12
len(lstItems):  11
before for-loop, len(lstItems) is:  10
len(lstItems):  10
len(lstItems):  9
len(lstItems):  8
before for-loop, len(lstItems) is:  7
le

IndexError: pop from empty list

In [174]:
lst = [1,2,3,4,5,6,7,8,9,10]

while lst:
    if len(lst) == 0:
        print("len(lst) in check",len(lst))
        break
    
    print("len(lst) before pop", len(lst))
    lst.pop()
    print("len(lst) after pop", len(lst))

print("lst after while loop", lst)

len(lst) before pop 10
len(lst) after pop 9
len(lst) before pop 9
len(lst) after pop 8
len(lst) before pop 8
len(lst) after pop 7
len(lst) before pop 7
len(lst) after pop 6
len(lst) before pop 6
len(lst) after pop 5
len(lst) before pop 5
len(lst) after pop 4
len(lst) before pop 4
len(lst) after pop 3
len(lst) before pop 3
len(lst) after pop 2
len(lst) before pop 2
len(lst) after pop 1
len(lst) before pop 1
len(lst) after pop 0
lst after while loop []


**Qn 3)** Write a `Circle` and a `Cylinder` class with the following requirements:

**`Circle` class**
* has 2 private variables `radius` & `colour`
* the `radius` & `colour` properties are set during initialization, if nothing is given, `radius` & `colour` defaults to `1.0` and `'Blue'` respectively
* setter method for the variable `radius` 
* accessor methods for the variable `colour`
* the `Cicrle` class must know how to calculate its own area
* the `Circle` class must return a string representation following the format `'Circle[radius=<?>, colour=<?>]'` when called from the `print()` function
* round all numbers to 2 decimal places

**`Cylinder` class**
* is derived from the `Circle` class
* has 1 additional private variable `height`
* the `height`, `radius` & `colour` properties are set during initialization, if nothing is given, `height`, `radius` & `colour` defaults to `1.0`, `1.0` and `'Red'` respectively
* accessor methods for the variable `height`
* the `Cylinder` class must know how to calculate its own volume
* the `Cylinder` class must return a string representation following the format `'Cylinder[radius=<?>, height=<?>, colour=<?>]'` when called from the `print()` function
* round all numbers to 2 decimal places

The test program have been provided.

**Sample Program Output**
<pre>
Cylinder 1 stats: Cylinder[radius=1.00, height=1.00, colour=Red]
Cylinder 1 volume: 3.14

Cylinder 2 stats: Cylinder[radius=1.00, height=10.00, colour=Red]
Cylinder 2 volume: 31.4

Cylinder 3 stats: Cylinder[radius=5.00, height=20.00, colour=black]
Cylinder 3 volume: 1570.8

Circle 1 stats: Circle[radius=4.00, colour=Blue]
Circle 1 area: 50.27
</pre>

In [35]:
import math

class Circle:
    
    def __init__(self, radius = 1.0, colour = 'Blue'):
        self.__radius = radius
        self.__colour = colour
        super().__init__()
      
    # Getters
    def getRadius(self):
        return self.__radius
    
    def getColour(self):
        return self.__colour
    
    # Setters
    def setColour(self, colour):
        
        if not isinstance(colour, str):
            raise TypeError("Colour is not of string type.")
            
        else:
            self.__colour = colour
    
    def setRadius(self, radius):
        
        if not isinstance(radius, (float, int)):
            raise TypeError("Radius is not of float or int type.")
            
        else:
            self.__radius = radius
            
    # Others            
    def cal_area(self):
        return math.pi*self.getRadius()**2
    
    def circumference(self):
        return math.pi*2*self.getRadius()
    
    def __str__(self):
        return f'Circle[Colour = {self.getColour()}, Radius= {self.getRadius():.2f}]'

In [38]:
class Cylinder(Circle):
    
    def __init__(self, height = 1.0, radius = 1.0, colour = 'Red'):
        super().__init__(radius, colour)
        self.__height = height
        
    def getHeight(self):
        return self.__height
    
    def setHeight(self, height):
        if not isinstance(height, (float, int)):
            raise TypeError("Height is not of float or int type.")
            
        else:
            self.__height = height
            
    def cal_volume(self):
        return self.getHeight()*(self.cal_area())
    
    def __str__(self):
        return f'Cylinder[radius={self.getRadius():.2f}, height= {self.getHeight():.2f}, colour={self.getColour()}]'
        

In [39]:
# Cylinder object 1 ---------------------------
cyl1 = Cylinder()
print(f'Cylinder 1 stats: {cyl1}')
print(f'Cylinder 1 volume: {cyl1.cal_volume()}\n')

# Cylinder object 2 ---------------------------
cyl2 = Cylinder(10.0)
print(f'Cylinder 2 stats: {cyl2}')
print(f'Cylinder 2 volume: {cyl2.cal_volume()}\n')

# Cylinder object 3 ---------------------------
cyl3 = Cylinder(20, 5, 'black')
print(f'Cylinder 3 stats: {cyl3}')
print(f'Cylinder 3 volume: {cyl3.cal_volume()}\n')

# Circle object 1 ---------------------------
cir1 = Circle(4)
print(f'Circle 1 stats: {cir1}')
print(f'Circle 1 area: {cir1.cal_area()}')

Cylinder 1 stats: Cylinder[radius=1.00, height= 1.00, colour=Red]
Cylinder 1 volume: 3.141592653589793

Cylinder 2 stats: Cylinder[radius=1.00, height= 10.00, colour=Red]
Cylinder 2 volume: 31.41592653589793

Cylinder 3 stats: Cylinder[radius=5.00, height= 20.00, colour=black]
Cylinder 3 volume: 1570.7963267948967

Circle 1 stats: Circle[Colour = Blue, Radius= 4.00]
Circle 1 area: 50.26548245743669


**Qn 4)** Write a `Shape`, a `Circle`, a `Rectangle` and a `Square` classes with the following requirements:

**`Shape` class**
* has 2 private variables `filled` (boolean) & `colour` (string)
* the `filled` & `colour` properties are set during initialization, if nothing is given, `filled` & `colour` defaults to `True` and `'Green'` respectively
* accessor methods for both variables
* the `Shape` class must return a string representation following the format `'Shape[colour=<?>, filled=<?>]'` when called from the `print()` function

**`Circle` class**
* is derived from the `Shape` class
* has 1 additional private variable `radius`
* its properties are set during initialization, if nothing is given, `radius` defaults to `1.0`.
* accessor methods for the `radius` variable
* the `Circle` class must know how to calculate its own area and perimeter
* the `Circle` class must return a string representation following the format `'Circle[Shape[filled=<?>, colour=<?>], radius=<?>]'` when called from the `print()` function
* round all numbers to 2 decimal places

**`Rectangle` class**
* is derived from the `Shape` class
* has 2 additional private variable `width` & `length`
* its properties are set during initialization, if nothing is given, both `width` & `length` defaults to `1.0`.
* accessor methods for both the `width` & `length` variable
* the `Rectangle` class must know how to calculate its own area and perimeter
* the `Circle` class must return a string representation following the format `'Rectangle[Shape[filled=<?>, colour=<?>], width=<?>, length=<?>]'` when called from the `print()` function
* round all numbers to 2 decimal places

**`Square` class**
* is derived from the `Rectangle` class
* has 1 additional private variable `side`
* its properties are set during initialization, if nothing is given, `side` defaults to `1.0`.
* accessor methods for the `side` (Note that the accessor methods should set both the `width` & `length` of the parent class and get the value of either `width` or `length` from the parent class)
* the `Square` class must return a string representation following the format `'Square[Rectangle[Shape[filled=<?>, colour=<?>], width=<?>, length=<?>]]'` when called from the `print()` function
* round all numbers to 2 decimal places

**Sample Program Output**
<pre>
Circle[Shape[colour=Green, filled=True], radius=5]
area=78.54
perimeter=31.42

Square[Rectangle[Shape[colour=Green, filled=True], width=5, length=5]]
area=25
perimeter=20
</pre>

In [43]:
class Shape:
    
    def __init__(self, filled = True, colour = 'Green'):
        self.__filled = filled
        self.__colour = colour
        
    def setColour(self, colour):
        
        if not isinstance(colour, str):
            raise TypeError("Colour is not of string type.")
            
        else:
            self.__colour = colour
            
    def setFilled(self, filled):
        
        if not isinstance(filled, bool):
            raise TypeError("Filled is not of bool type.")
            
        else:
            self.__filled = filled
            
    def getFilled(self):
        return self.__filled
            
    def getColour(self):
        return self.__colour
        
    def __str__(self):
        return f"Shape[Colour = {self.getColour()}, Filled = {self.getFilled()}]"

**`Circle` class**
* is derived from the `Shape` class
* has 1 additional private variable `radius`
* its properties are set during initialization, if nothing is given, `radius` defaults to `1.0`.
* accessor methods for the `radius` variable
* the `Circle` class must know how to calculate its own area and perimeter
* the `Circle` class must return a string representation following the format `'Circle[Shape[filled=<?>, colour=<?>], radius=<?>]'` when called from the `print()` function
* round all numbers to 2 decimal places

In [42]:
import math

class Circle(Shape):
    
    def __init__(self, radius = 1.0):
        self.__radius = radius
        super().__init__()
        
    def getRadius(self):
        return self.__radius
    
    def setRadius(self, radius):
        
        if not isinstance(radius, (float, int)):
            raise TypeError("Radius is not of float or int type.")
            
        else:
            self.__radius = radius
            
    def area(self):
        return math.pi*self.getRadius()**2
    
    def circumference(self):
        return math.pi*2*self.getRadius()
    
    def __str__(self):
        sp = super().__str__()
        return f'Circle[{sp}, Radius= {self.getRadius():.2f}]'

**`Rectangle` class**
* is derived from the `Shape` class
* has 2 additional private variable `width` & `length`
* its properties are set during initialization, if nothing is given, both `width` & `length` defaults to `1.0`.
* accessor methods for both the `width` & `length` variable
* the `Rectangle` class must know how to calculate its own area and perimeter
* the `Circle` class must return a string representation following the format `'Rectangle[Shape[filled=<?>, colour=<?>], width=<?>, length=<?>]'` when called from the `print()` function
* round all numbers to 2 decimal places


In [41]:
class Rectangle(Shape):
    
    def __init__(self, width = 1.0, length = 1.0):
        super().__init__()
        self.__width = width
        self.__length = length
        
    def getWidth(self):
        return self.__width
    
    def getLength(self):
        return self.__length
    
    def setWidth(self, width):
        
        if isinstance(width, (float, int)):
            raise TypeError("Width is not of float or int type.")
            
        else:
            self.__width = width
            
    def setLength(self, length):

        if isinstance(length, (float, int)):
            raise TypeError("length is not of float or int type.")

        else:
            self.__length = length
            
    def area(self):
        return self.getLength()*self.getWidth()
    
    def perimeter(self):
        return 2*(self.getLength()+self.getWidth())
    
    def __str__(self):
        sp = super().__str__()
        return f'Rectangle[{sp}, width={self.getWidth():.2f}, length={self.getLength():.2f}]'


**`Square` class**
* is derived from the `Rectangle` class
* has 1 additional private variable `side`
* its properties are set during initialization, if nothing is given, `side` defaults to `1.0`.
* accessor methods for the `side` (Note that the accessor methods should set both the `width` & `length` of the parent class and get the value of either `width` or `length` from the parent class)
* the `Square` class must return a string representation following the format `'Square[Rectangle[Shape[filled=<?>, colour=<?>], width=<?>, length=<?>]]'` when called from the `print()` function
* round all numbers to 2 decimal places

In [45]:
class Square(Rectangle):
    
    def __init__(self, side = 1.0):
        super().__init__(side, side) # Need to pass in variables.
        self.__side = side
        
    def getSide(self):
        return super().getLength()
    
    def setSide(self, side):
        
        if isinstance(side, (float, int)):
            raise TypeError("Side is not of type int or float.")
            
        else:
            super().getWidth(side)
            super().getLength(side)
            self._side = side
            
    def __str__(self):
        sp = super().__str__()
        return f'Square[{sp}, Width= {self.getSide():.2f}, Length= {self.getSide():.2f}]]'

In [46]:
c = Circle(5)
print(c)
print(f'area={c.area()}')
print(f'perimeter={c.circumference()}')

print()
s = Square(5)
print(s)
print(f'area={s.area()}')
print(f'perimeter={s.perimeter()}')

Circle[Shape[Colour = Green, Filled = True], Radius= 5.00]
area=78.53981633974483
perimeter=31.41592653589793

Square[Rectangle[Shape[Colour = Green, Filled = True], width=5.00, length=5.00], Width= 5.00, Length= 5.00]]
area=25
perimeter=20
